internal/client/service.go

```diff
@@ -0,0 +1,31 @@
+package client
+
+import (
+	"grabvn-golang-bootcamp/internal/common"
+)
+
+type EchoClient interface {
+	GetEcho() (common.HttpResponse, error)
+}
+
+type EchoClientMiddleware func(EchoClient) EchoClient
+
+type echoClient struct {
+	url        string
+	httpClient common.HttpClient
+}
+
+func NewEchoClient(url string) *echoClient {
+	return &echoClient{url: url, httpClient: common.NewHttpClient()}
+}
+
+func (e *echoClient) GetEcho() (common.HttpResponse, error) {
+	res, err := e.httpClient.Get(e.url)
+	if err != nil {
```

Comment:
do we retry for every kinds error? `500 <= res.GetStatusCode` should be included in this `if`